### ISSUE TITLE
Iterable type error not caught

### ISSUE THREAD

**Issue by desmond-dsouza** [JukkaL/mypy#1036]
Here is the code, the type error in `s_cross` (`for y in l:` should be `for y in l():`) is not caught by mypy, resulting in a run-time TypeError:

```
from typing import TypeVar, Callable, Iterable

T = TypeVar("T")

def identity(x: T) -> T:
    return x

def s_map(f: Callable[[T], T], l: Iterable[T]) -> Callable[[], Iterable[T]]:
    return lambda: map(f, l)

def s_cross(l1):
    l = s_map(identity, l1)
    for x in l():
        for y in l:
            yield x, y

L = [1, 2, 3, 4]

print(list(s_cross(L)))
```

**Comment by JukkaL** [comment]
Mypy doesn't type check functions without annotations. Does mypy catch the error if you add an annotation to `s_cross`?

**Comment by desmond-dsouza** [comment]
Yes, it does catch the error with an annotated `s_cross`.

Wouldn't type checking of non-annotated functions be useful, with the usual `Any` defaults? Wouldn't It would still catch things like this error?

Thanks.

**Comment by JukkaL** [comment]
Yeah, we are planning to support that in the future (it's currently called 'weak mode'), but we need to be careful to not generate a ton of bogus warnings about unannotated functions, and that is pretty hard to do in general. But we are working on it :-)

**Comment by JukkaL** [comment]
Closing this as mypy is behaving as designed (though it's not optimal). 

I noticed that there is no issue for weak mode so I created one: #1038. Feel free to continue discussion there.
